File: TypeStrictTests/Interface_ElX.ts
module tsp {

    export interface IDOMBinder {

        attributes?: { [name: string]: string; };
        contentEditable?: bool;
        dynamicAttributes?: { [name: string]: { (el: IElX): string; }; };
        dynamicClasses?: { [name: string]: { (el: IElX): bool; }; };
        dynamicStyles?: { [name: string]: { (el: IElX): string; }; };
        b;
        classes?: string[];

        id?: string;

        kids?: IRenderable[];
        kidsGet?: (el: IElX)=> IElX[];

        styles?: { [name: string]: string; };

        tag?: string;
        //Inner Content - used if textGet is null
        text?: string;
        //Dynamic Inner Content
        textGet? (el: IElX): string;
        //child elements - used if kidsGet is null

        toggleKidsOnParentClick?: bool;
        collapsed?: bool;
        dataContext?: any;
        selectSettings?: ISelectBinder;
        container?: any;
        onNotifyAddedToDom?: (el: IElX) => any;
    }

    export interface IRenderable {
        parentElement: IRenderable;
        doRender(context: IRenderContext);
        ID: string;
    }

    export interface IElX extends IRenderable {
        bindInfo: IDOMBinder;
        //parentElement: IElX;
        //doRender(context: IRenderContext);
        //ID: string;
        el: HTMLElement;
        
        kidElements: IElX[];
        selected: bool;
        _rendered: bool;
        innerRender(settings: IRenderContextProps);
        removeClass(className: string);
        ensureClass(className: string);
    }

    export interface IRenderContext {
        output: string;
        elements: IElX[];
        settings: IRenderContextProps;
    }

    export interface ISelectBinder {
        //static 
        selected?: bool;
        //dynamic
        selectGet? (elX: IElX): bool;
        selectSet? (elX: IElX, newVal: bool): void;
        group?: string;
        selClassName?: string;
        partialSelClassName?: string;
        unselClassName?: string;
        conformWithParent?: bool;
    }

    export interface IListenForTopic {
        topicName: string;
        conditionForNotification? (tEvent: ITopicEvent): bool;
        callback(tEvent: ITopicEvent): void;
        /**
        element extension listening for the event
        */
        elX?: IElX;
        elXID?: string;
    }

    export interface ITopicEvent extends IListenForTopic {
        event: Event;
    }

    export interface IRenderContextProps {
        targetDomID?: string;
        targetDom?: HTMLElement;
    }
}

File: TypeStrictTests/Interface_ElX.cs
using System;
using System.Collections.Generic;
//using CurlyBraceParser.DOM;

public static partial class tsp
{
    public interface IDOMBinder
    {
        Dictionary<string, string> attributes { get; set; }
        bool contentEditable { get; set; }
        Dictionary<string, Func<IElX, string>> dynamicAttributes { get; set; }
        Dictionary<string, Func<IElX, bool>> dynamicClasses { get; set; }
        Dictionary<string, Func<IElX, string>> dynamicStyles { get; set; }
        string[] classes { get; set; }
        string id { get; set; }
        IRenderable[] kids { get; set; }
        IElX[] kidsGet(IElX el);
        Dictionary<string, string> styles { get; set; }
        string tag { get; set; }
        string text { get; set; }
        string textGet(IElX el);
        bool toggleKidsOnParentClick { get; set; }
        bool collapsed { get; set; }
        object dataContext { get; set; }
        ISelectBinder selectSettings { get; set; }
        object container { get; set; }
        object onNotifyAddedToDom(IElX el);
    }
    public interface IRenderable
    {
        IRenderable parentElement { get; set; }
        object doRender(IRenderContext context  = null);
        string ID { get; set; }
    }
    public interface IElX
    {
        IDOMBinder bindInfo { get; set; }
        //HTMLElement el { get; set; }
        IElX[] kidElements { get; set; }
        bool selected { get; set; }
        bool _rendered { get; set; }
        object innerRender(IRenderContextProps settings);
        object removeClass(string className);
        object ensureClass(string className);
    }
    public interface IRenderContext
    {
        string output { get; set; }
        IElX[] elements { get; set; }
        IRenderContextProps settings { get; set; }
    }
    public interface ISelectBinder
    {
        bool selected { get; set; }
        bool selectGet(IElX elX);
        void selectSet(IElX elX, bool newVal);
        string group { get; set; }
        string selClassName { get; set; }
        string partialSelClassName { get; set; }
        string unselClassName { get; set; }
        bool conformWithParent { get; set; }
    }
    public interface IListenForTopic
    {
        string topicName { get; set; }
        bool conditionForNotification(ITopicEvent tEvent);
        void callback(ITopicEvent tEvent);
        IElX elX { get; set; }
        string elXID { get; set; }
    }
    public interface ITopicEvent
    {
        //Event topicEvent { get; set; }
    }
    public interface IRenderContextProps
    {
        string targetDomID { get; set; }
        //HTMLElement targetDom { get; set; }
    }
}
